draft-bbs-signatures.md

```diff
@@ -291,6 +291,63 @@ For readability, this document makes these transformations implicitly, but they
 
 Optional input/parameters to operations that feature in a call to a HASH or XOF function, that are not supplied to the operation should default to an empty octet string. For example, if X is an optional input/parameter that is not supplied, whilst A and B are required, then the procedural step of `HASH(A || X || B)` MUST be evaluated to `HASH(A || "" || B)`.
 
+## Generator point computation
+
+The `compute_generators` function defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify. It takes one input, the maximum number of selectively revealed messages, which determines the number of generator points to create. The result is a sequence beginning with the `H_s` and `H_d` generators, followed by each of the `H_i` message generators.
+
+This document defines one implementation of this method, `compute_generators_xof`, which uses [@!I-D.irtf-cfrg-hash-to-curve] with an extensible-output hash function to generate a sequence of points with a uniformly random distribution from a seed value and DST.
+
+Each generator creation method must define a unique `GENERATORS_ID` identifier which is incorporated into the ciphersuite ID.
+
+### Security Considerations
+
+TODO
+
+### compute\_generators\_xof
+
+```
+generators = compute_generators_xof(maxMessageCount);
+
+Parameters:
+
+- hash_to_curve_g1: A cryptographic hash function that takes as an arbitrary octet string input and returns a point in G1 as defined in [@!I-D.irtf-cfrg-hash-to-curve]. The expand_message method MUST be expand_message_xof, and hash MUST be the same as xof.
+- generator_xof, an extensible-output cryptographic hash function.
+- generator_dst, octet string. The domain separation tag.
+- generator_seed, octet string. A unique seed value for the generator point sequence.
+
+Inputs:
+
+- maxMessageCount (REQUIRED), non-negative integer. The maximum number of selectively-revealed messages to be supported.
+
+Outputs:
+
+- generators, an array of generator points of length `maxMessageCount + 2`.
+
+Procedure:
+
+1. h = generator_xof(generator_seed)
+
+2. length = maxMessageCount + 2
+
+3. for i in 1 to length:
+
+4.    generator_i = Identity_G1
+
+5.    while(generator_i == Identity_G1 or generator_i == P1)
+
+6.        candidate = hash_to_curve_g1(h.read(xof_no_of_bytes), generator_dst)
+
+7.        if candidate not in generators: generator_i = candidate
+
+8. return generators
+```
+
```

Comment:
Since we are making the update, should we maybe consider a different procedure as well?? Something like,
```
Procedure:

1. L= maxMessageCount + 2
2. h_0 = hash(seed || I2OSP(0, 2))
3. h_1 = hash(h_0 || I2OSP(1, 2))
4. for i in 1 to L:
5.    generator_i = Identity_G1
6.    while(generator_i == Identity_G1 or generator_i == P1)
7.        candidate = hash_to_curve_g1(h_i, generator_dst)
8.        h_i = hash(strxor(h_0, h_(i-1)) || I2OSP(i, 2))
9.        if candidate not in generators: generator_i = candidate
10. return generators
```
Here `hash` can be an xof also. What i like about this is that let's say you have cached 1000 generators and you need to sign 1001 messages. In the current procedure you need to compute the 1000 generators from the beginning.  In this approach though, You only need to save the last `h_i` computed, and immediately calculate the next generator. Some notes,

- The input seed is not adversarially controlled so no need for more complicated procedures (because of the guarantees that this function gives, i.e., close to uniform output distribution)
- Is less efficient, but i really don't think that time efficiency is what we should optimize for here. Also, cases like the above will hurt performance. 
- Is usable with whatever hash is defined by the ciphersuite, so no need for extra parameters (this is related with the discussion in #162 as well)

Thoughts?? c.c. @mikelodder7, @tplooker^^

Reply:
I like the idea of using expand_message rather than defining something equivalent, but this method does make it harder to extend a cached list of generators.

Reply:
Good point. Using `expand_message_xmd` you could save `b_0, last_i, b_(last_i)` as state and start from where you left of (and it will be easy to describe in the spec IMO). Using `expand_message_xof` though will be harder. However the current xof based construction has the same issue. So if people want this ability they can always use `expand_message_xmd` with an xof??